kessel/auth/auth.go

```diff
@@ -0,0 +1,150 @@
+package auth
+
+import (
+	"context"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/zitadel/oidc/v3/pkg/client"
+)
+
+const expirationWindow = 300  // 5 minutes in second
+const defaultExpiresIn = 3600 // 1 hour in seconds
+
+type OIDCDiscoveryMetadata struct {
+	TokenEndpoint string
+}
+
+type RefreshTokenResponse struct {
+	AccessToken string
+	ExpiresAt   time.Time
+}
+
+type OAuth2ClientCredentials struct {
+	clientId      string
+	clientSecret  string
+	tokenEndpoint string
+	cachedToken   RefreshTokenResponse
+	tokenMutex    sync.Mutex
+}
+
+type FetchOIDCDiscoveryOptions struct {
+	Context    context.Context
+	HttpClient *http.Client
+}
+
+type GetTokenOptions struct {
+	ForceRefresh bool
+	Context      context.Context
+	HttpClient   *http.Client
+}
+
+type oauth2TokenEndpointCaller struct {
+	tokenEndpoint string
+	httpClient    *http.Client
+}
+
+type requestToken struct {
+	ClientID     string `schema:"client_id,omitempty"`
+	ClientSecret string `schema:"client_secret,omitempty"`
+	GrantType    string `schema:"grant_type"`
+}
+
+func MakeOAuth2ClientCredentials(clientId string, clientSecret string, tokenEndpoint string) OAuth2ClientCredentials {
```

Comment:
Curious why we are using the Make prefix for the constructor here

Reply:
I was under the impression that `Make` prefix was used to create non-pointer structs, but I guess i should be using `New` instead?